[#55][news] 뉴스 단건 조회, 복수 조회, 검색 구현

luz315 opened the pull request:
## Description
- 쿼리 DSL 설정 
- 도메인 로직 수정
  - base 엔티티 주입
  - 외부 라이브러리 안쓰고 커스텀 페이지네이션 구현
  - 검색 조건 디티오 구현
  - 리스트형태 디티오 구현
- 뉴스 단건조회, 뉴스 복수 조회 query dsl 로직 구현
  - 단건 조회 fetchjoin 사용
  - 복수 조회 2단계 전략을 통해 fetchjoin 사용
  - 페이징 처리완료
  - 조건부 검색 완료
- 컨슈머 dlq로직 보완
- 권한 추가

Resolves: #55

## PR Type
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

luz315 committed:
[#55][news] refactor: 쿼리 DSL 설정 및 불필요 로직 제거

luz315 committed:
[#55][news] feat: 도메인 로직 수정

- base 엔티티 주입
- 외부 라이브러리 안쓰고 커스텀 페이지네이션 구현
- 검색 조건 디티오 구현
- 리스트형태 디티오 구현

luz315 committed:
[#55][news] feat: 뉴스 단건조회, 뉴스 복수 조회 query dsl 로직 구현

- 단건 조회 fetchjoin 사용
- 복수 조회 2단계 전략을 통해 fetchjoin 사용
- 페이징 처리완료
- 조건부 검색 완료

luz315 committed:
[#55][news] feat: 컨슈머 dlq로직 보완

luz315 committed:
[#55][news] feat: 권한 추가

hp5234 commented on news/src/main/java/com/newpick4u/news/news/application/dto/response/NewsListResponse.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.UUID;
+
+public record NewsListResponse(
+        int page,
```
Common 서비스 에 PageResponse<T> 를 사용해도 되는지 검토 해봐도 될것같습니다.

luz315 replied:
해당부분 리팩토링 해놔서 
- news - domain - model - Pagination.java
- news - application - dto - PageResponse.java
- news - application - dto - NewsSummaryDto.java

적용은 컨트롤러 서비스 로직에서 searchNewsList 메서드 반환값을 참고하시면 됩니다~!

hp5234 commented on news/src/main/java/com/newpick4u/news/news/infrastructure/jpa/NewsRepositoryAdapter.java:
```diff
@@ -0,0 +1,60 @@
+
+@Repository
+@RequiredArgsConstructor
+public class NewsRepositoryAdapter implements NewsRepository {
```
리파지토리들 네이밍을 통일해도 괜찮을것같습니다. 
저는 아래 방식을 사용하고 있고 해당 팀원들과 이야기해도 괜찮을거같아요. 의견 주세요.

NewsRepositoryAdapter -> NewsRepositoryImpl
JpaNewsRepository -> NewsJpaRepository

luz315 replied:
NewsRepositoryAdapter -> NewsRepositoryImpl 이거는 말씀해주신 걸로 바꾸는게 더명확할거 같습니다!  NewsRepositoryImpl 이걸로 가도록하고 jpa는 회의하면 좋을 것 같습니다

hp5234 commented on news/src/main/java/com/newpick4u/news/news/infrastructure/jpa/NewsRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public Pagination<News> searchNewsList(NewsSearchCriteria request, boolean isMaster) {
+        QNews news = QNews.news;
```
Q객체는 보통 static import 로 사용하는 편입니다. 이것도 의견 주세요

luz315 replied:
수정했습니다~

hp5234 commented on news/src/main/java/com/newpick4u/news/news/presentation/NewsController.java:
```diff
@@ -0,0 +1,35 @@
+    private final NewsService newsService;
+
+    @GetMapping("/{id}")
+    public ResponseEntity<NewsResponseDto> getNews(
```
응답 객체를 CommonResponse 로 한번 래핑해야할것으로 보입니다.

luz315 replied:
커먼 리스폰스 좀 더 깔끔하게 리팩토링 하면 좋을 것 같아서 잠시 빼두었습니다 :)

hp5234 commented on news/src/main/java/com/newpick4u/news/news/presentation/NewsController.java:
```diff
@@ -0,0 +1,35 @@
+@RestController
+@RequestMapping("api/v1/news")
+@RequiredArgsConstructor
+public class NewsController {
```
컨벤션에 의해 NewsApiController 가 되어야할것같습니다.

luz315 replied:
수정했습니다~

venus-y commented on news/src/main/java/com/newpick4u/news/news/application/usecase/NewsServiceImpl.java:
```diff
@@ -71,4 +78,25 @@ private String serializeToJson(NewsTagDto dto) {
+        if (userInfoDto.role() == UserRole.ROLE_MASTER) {
+            news = newsRepository.findDetail(id)
+                    .orElseThrow(() -> new IllegalArgumentException("뉴스를 찾을 수 없습니다."));
+        } else {
```
if문 마지막 줄에서
` return NewsResponseDto.from(news);` 과 같이 구성하면 else 블럭을 생략할 수 있을 것 같은데 어떻게 생각하시는지 궁금합니다.

luz315 replied:
해당부분 쿼리 dsl적용해 고쳐놓았습니다

luz315 committed:
[#55][news] feat: 뉴스 단건조회 리팩토링

- 뉴스 단건조회 querydsl 적용
- querydsl extract method 적용
- if-else 분기처리 로직 제거
- 리포지토리명 컨벤션 적용
- 불필요로직 제거

luz315 committed:
[#55][news] refactor: 페이징 리팩토링

- 페이지네이션 커스텀 모델 기능추가
- 페이지 리스폰스 구현

luz315 committed:
[#55][news] refactor: 뉴스 필드 추가

- url, publishedDate 추가

luz315 committed:
[#55][news] refactor: 컨슈머 예외 중복 제거

- url, publishedDate 추가

dannyseo9202 commented on news/src/main/java/com/newpick4u/news/news/application/dto/response/NewsListResponse.java:
```diff
@@ -0,0 +1,52 @@
+        long totalPages,
+        List<NewsSummary> contents
+) {
+    public static NewsListResponse from(Pagination<News> pagination) {
```
👍 Pagination 부분 잘 구현하신 것 같습니다 소라님. 
도메인 계층, 리포지토리에서 org.springframework.data.domain.Page; 사용하지 않으신 점이 좋은 것 같습니다 ~!

dannyseo9202 commented on news/src/main/java/com/newpick4u/news/news/domain/critria/NewsSearchCriteria.java:
```diff
@@ -0,0 +1,10 @@
+package com.newpick4u.news.news.domain.critria;
+
+public record NewsSearchCriteria(
```
👍

luz315 committed:
[#52][news] refactor: api 수정

dannyseo9202 commented on news/src/main/java/com/newpick4u/news/news/domain/entity/News.java:
```diff
@@ -26,6 +27,12 @@ public class News {
+    private String url;
+
+    @Column(nullable = false)
+    private String publishedDate;
```
소라님, 해당 변수명이 ~Date로 끝나는 것으로 보아 시간 기록 용도로 사용하시려는 의도로 보입니다.
혹시 시간을 String 타입으로 저장하고 계시다면, LocalDate나 LocalDateTime 같은 시간 타입 사용도 고려해보시면 좋을 것 같아요 :) (_ _)

hp5234 reviewed (approved):
고생하셨습니다.

dannyseo9202 commented on news/src/main/java/com/newpick4u/news/news/domain/model/Pagination.java:
```diff
@@ -0,0 +1,28 @@
+
+@Getter
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class Pagination<T> {
```
해당 부분 정말 잘 구현하신 것 같습니다!
특히 도메인 계층에서 인프라 성격의 요소 (스프링 프레임 워크 관련된 요소)
ex) org.springframework.data.domain.Page

들을 도메인 계층에서 분리해주신 점이 인상 깊은 것 같습니다.
의도와 구현이 잘 맞아떨어지는 느낌입니다 👍

dannyseo9202 commented on news/src/main/java/com/newpick4u/news/news/infrastructure/jpa/NewsRepositoryimpl.java:
```diff
@@ -0,0 +1,47 @@
+
+@Repository
+@RequiredArgsConstructor
+public class NewsRepositoryimpl implements NewsRepository {
```
👍 목적에 맞게 위임 처리를 잘 해주신 것 같습니다!
NewsRepositoryImpl도 JPA와 QueryDSL을 함께 사용하기 위한 클래스로서의 역할을 잘 수행하고 있는 것 같습니다.

dannyseo9202 commented on news/src/main/java/com/newpick4u/news/news/infrastructure/kafka/RetryNewsTagConsumer.java:
```diff
@@ -20,14 +23,26 @@ public class RetryNewsTagConsumer {
             containerFactory = "newsTagListenerContainerFactory"
     )
-    public void consume(ConsumerRecord<String, NewsTagDto> record) {
+    public void consume(ConsumerRecord<String, NewsTagDto> record, Acknowledgment ack) {
```
소라님 
1. byte 배열에서 바로 value[0]을 꺼내 int로 캐스팅하고 있는데, 이는 음수로 변환되거나 의도치 않은 값이 나올 수 있습니다. Java의 ByteBuffer를 사용하는 것이 더 안전합니다.

2. attempt >= 3 및 RETRY_HEADER_KEY는 외부 상수로 추출하는 것이 유지보수에 좋습니다.



```
@Slf4j
@Component
@RequiredArgsConstructor
public class RetryNewsInfoConsumer {

    private static final int MAX_RETRY_ATTEMPTS = 3;
    private static final String RETRY_HEADER_KEY = "kafka_NewsAttempt";

    private final NewsService newsService;

    @KafkaListener(
            topics = "news-info-dlq.fct.v1",
            groupId = "news-info-dlq-consumer",
            containerFactory = "newsTagListenerContainerFactory"
    )
    public void consume(ConsumerRecord<String, NewsInfoDto> record, Acknowledgment ack) {
        NewsInfoDto dto = record.value();
        int attempt = getRetryAttempt(record);

        log.info("[DLQ Retry] 뉴스 초안 DLQ 재처리 {}회차 시작 - {}", attempt, dto);

        try {
            newsService.saveNewsInfo(dto);
            log.info("[DLQ Retry] 뉴스 초안 DLQ 재처리 성공 - {}", dto);
            ack.acknowledge();
        } catch (Exception e) {
            log.error("[DLQ Retry] 뉴스 초안 재처리 {}회차 실패 - {}", attempt, dto, e);
            if (attempt >= MAX_RETRY_ATTEMPTS) {
                log.error("[DLQ Retry] 최대 재시도 초과. 메시지 폐기 - {}", dto);
                ack.acknowledge(); // 폐기하고 커밋
            }
        }
    }

    private int getRetryAttempt(ConsumerRecord<?, ?> record) {
        return Optional.ofNullable(record.headers().lastHeader(RETRY_HEADER_KEY))
                .map(Header::value)
                .map(ByteBuffer::wrap)
                .map(ByteBuffer::getInt)
                .orElse(1);
    }
}

```

venus-y reviewed (approved):
고생하셨습니다 ㅎㅎ

dannyseo9202 commented on news/src/main/java/com/newpick4u/news/news/presentation/NewsAPIController.java:
```diff
@@ -0,0 +1,37 @@
+import com.newpick4u.news.news.application.dto.response.NewsSummaryDto;
+import com.newpick4u.news.news.application.dto.response.PageResponse;
+import com.newpick4u.news.news.application.usecase.NewsService;
+import com.newpick4u.news.news.domain.critria.NewsSearchCriteria;
```
소라님 해당 부분 도메인 계층의 NewsSearchCriteria를 직접 참조하지 않게 별도의 DTO로 분리 부탁드릴게요 ~!

dannyseo9202 reviewed (commented):
소라님, 리뷰 완료했습니다!
전반적으로 계층 분리를 정말 잘 해주신 것 같아요.
4계층 구조와 DIP에 대한 이해도가 확실히 느껴지는 구현이었습니다 ㅎㅎ
수고 많으셨어요~!! (_ _)
